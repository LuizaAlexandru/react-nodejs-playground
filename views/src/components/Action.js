import React, { Component, PropTypes } from 'react';

export default class Action extends Component {
  constructor(props) {
    super(props);
    this.handleClickAction = this.handleClickAction.bind(this);
  }

  render() {
    return <div className="action" onClick={this.handleClickAction}>{
      this.props.job == "remove" ?
      <span className="action-placeholder">{this.props.placeholder}</span> : this.props.text
    }</div>
  }

  handleClickAction(e) {
    e.preventDefault();
    const props = this.props,
      productId = props.id,
      job = props.job;

    if (job == 'add') {
      this.props.addProduct(productId);
    } else if (job == 'remove') {
      this.props.removeProduct(productId);
    }
  }
}



// Action.propTypes = {
//   text: PropTypes.string
// }
