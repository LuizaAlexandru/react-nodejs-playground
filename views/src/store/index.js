import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/index";

const asyncDispatchMiddleware = store => next => action => {
  let syncActivityFinished = false;
  let actionQueue = [];

  function flushQueue() {
    actionQueue.forEach(a => store.dispatch(a)); // flush queue
    actionQueue = [];
  }

  function asyncDispatch(asyncAction) {
    actionQueue = actionQueue.concat([asyncAction]);

    if (syncActivityFinished) {
      flushQueue();
    }
  }

  const actionWithAsyncDispatch =
      Object.assign({}, action, { asyncDispatch });

  next(actionWithAsyncDispatch);
  syncActivityFinished = true;
  flushQueue();
};

const store = createStore(
  rootReducer,
  applyMiddleware(asyncDispatchMiddleware)
);

export default store;
