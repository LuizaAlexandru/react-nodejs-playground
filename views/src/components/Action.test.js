import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';

import Action from './Action.js';
import store from '../store/index';


test('Initiated Action matches the snapshot', () => {
  const component = renderer.create(
    <Provider store={store}>
      <Action />
    </Provider>
  );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('action has text', () => {
  let wrapper = mount(
    <Provider store={store}>
      <Action text="text property" />
    </Provider>);
    expect(wrapper.text()).toBe('text property');
});

test('handleClickAction preventsDefault event', () => {
  let wrapper = shallow(
      <Action text="text property" store={store} />
    );

    const action = wrapper.dive(),
      e = {};
    let flag = false;
      e.preventDefault = () => {
        flag = true;
      };
    action.simulate('click', e);
    expect(flag).toBe(true);
})
