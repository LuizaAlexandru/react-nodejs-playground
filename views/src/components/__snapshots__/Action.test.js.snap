// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initiated Action matches the snapshot 1`] = `
<div
  className="action"
  onClick={[Function]}
/>
`;
